### PR TITLE
Create cursor api

### PR DESCRIPTION
**Pull request by Pham-Gia-Huong**

### CODE REVIEW

File: app_test.go [hunk 1]
```diff
@@ -171,6 +172,32 @@ func TestDeleteRecord(t *testing.T) {
...
+}
+func TestCreateCurSor(t *testing.T) {
+	app := newAppWithApiToken(18)
+	result, err := app.createCursor([]string{"$id", "date"})
+	if err != nil {
+		t.Errorf("error %v", err)
+	}
+	fmt.Println("result", string(result))
+}
+
```
Comment (josh-vo) [inline app_test.go]: Please re-write unit test follow flow create cursor -> get/delete cursor.

File: app.go [hunk 2]
```diff
@@ -69,6 +91,26 @@ func (f UpdateKey) MarshalJSON() ([]byte, error) {
 	})
 }
 
+func (app *App) getCurSor(idCursor string) ([]byte, error) {
+	type requestBody struct {
+		Id string `json:"id,string"`
+	}
+
+	req, err := app.newRequest("GET", "records/cursor", nil, "id="+idCursor)
```
Comment (josh-vo) [inline app.go]: - "requestBody" variable isn't used.
- Query should be included in body.

File: app.go [hunk 3]
```diff
@@ -69,6 +91,26 @@ func (f UpdateKey) MarshalJSON() ([]byte, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	resGetCursor, err := app.do(req)
```
Comment (josh-vo) [inline app.go]: Could you please describe variable name more clearly? 
And could you please also do it with others, if variable name is conflicted with golang syntax you can skip it. 
Thank you.

File: app.go [hunk 4]
```diff
@@ -160,7 +202,7 @@ func (app *App) GetUserAgentHeader() string {
 	return app.userAgentHeader
 }
 
-func (app *App) newRequest(method, api string, body io.Reader) (*http.Request, error) {
+func (app *App) newRequest(method, api string, body io.Reader, query string) (*http.Request, error) {
```
[the remark this answers is not in the document]
Comment (Pham-Gia-Huong) [reply app.go]: The body is not work with query.

File: app.go (hunk 2 above)
Comment (Pham-Gia-Huong) [reply app.go]: body query is not work with get method for the new request

File: app_test.go [hunk 5]
```diff
@@ -171,6 +172,71 @@ func TestDeleteRecord(t *testing.T) {
 	}
 }
 
+func TestGetCursor(t *testing.T) {
+	app := newApp(18)
+
+	resultCreateCursor, err := app.createCursor([]string{"$id", "date"}, "", 100)
+	if err != nil {
+		t.Errorf("create cursor is fail: %v", err)
```
Comment (josh-vo) [inline app_test.go]: Could you please update the error messages with correct grammar?

File: app_test.go [hunk 6]
```diff
@@ -171,6 +172,63 @@ func TestDeleteRecord(t *testing.T) {
...
+	app := newApp(18)
+	if len(app.Password) == 0 {
+		t.Skip()
+	}
+
+	id := app.createCursorForTest()
+	result, err := app.deleteCursor(string(id))
+
+	if err != nil {
+		t.Errorf("delete cursor is fail: %v", err)
```
Comment (josh-vo) [inline app_test.go]: Please update the error message.

File: app_test.go [hunk 7]
```diff
@@ -171,6 +172,63 @@ func TestDeleteRecord(t *testing.T) {
...
+
+func TestCreateCurSor(t *testing.T) {
+
+	app := newAppWithApiToken(18)
+	if len(app.ApiToken) == 0 {
+		t.Skip()
+	}
+	_, err := app.createCursor([]string{"$id", "date"}, "", 100)
+	if err != nil {
+		t.Errorf("create cursor is fail: %v", err)
```
Comment (josh-vo) [inline app_test.go]: Please also update the error message.

File: app_test.go [hunk 8]
```diff
@@ -171,6 +172,63 @@ func TestDeleteRecord(t *testing.T) {
...
+	result, err := app.deleteCursor(string(id))
+
+	if err != nil {
+		t.Errorf("delete cursor is fail: %v", err)
+	}
+	fmt.Println(result)
+}
+
+func TestCreateCurSor(t *testing.T) {
+
```
Comment (josh-vo) [inline app_test.go]: Please remove the blank line.

File: app.go [hunk 9]
```diff
@@ -1000,3 +1053,76 @@ func (app *App) Fields() (map[string]*FieldInfo, error) {
...
+	result, err := parseResponse(response)
+	if err != nil {
+		return "", err
+	}
+	if result != nil {
+	}
+	return "delete success", nil
+}
+
+func (app *App) getCurSor(id string) ([]byte, error) {
```
Comment (josh-vo) [inline app.go]: Could you please correct the name of function?

File: app.go [hunk 10]
```diff
@@ -1000,3 +1053,76 @@ func (app *App) Fields() (map[string]*FieldInfo, error) {
...
+		return "", err
+	}
+
+	result, err := parseResponse(response)
+	if err != nil {
+		return "", err
+	}
+	if result != nil {
+	}
+	return "delete success", nil
```
Comment (josh-vo) [inline app.go]: Please remove the returned string, this function only return if error occurred. Please refer to DeleteRecords function.

File: app_test.go [hunk 11]
```diff
@@ -171,6 +172,63 @@ func TestDeleteRecord(t *testing.T) {
...
 
+func TestGetCursor(t *testing.T) {
+	app := newApp(18)
+
+	id := app.createCursorForTest()
+	result, err := app.getCurSor(string(id))
+
+	if len(app.Password) == 0 {
+		t.Skip()
+	}
```
Comment (josh-vo) [inline app_test.go]:
```
if len(app.Password) == 0 {
 	t.Skip()
 }
``` 
This logic should happen before creating cursor.

File: app.go [hunk 12]
```diff
@@ -1000,3 +1053,76 @@ func (app *App) Fields() (map[string]*FieldInfo, error) {
...
+	}
+	data := cursor{App: app.AppId, Fields: fields, Size: size, Query: query}
+	jsonData, _ := json.Marshal(data)
+	url := app.createUrl("records/cursor", "")
+	request, err := app.NewRequest("POST", url.String(), bytes.NewBuffer(jsonData))
+	if err != nil {
+		return nil, err
+	}
+	response, err := app.do(request)
+
```
Comment (josh-vo) [inline app.go]: Could you please remove this blank line, and using blank line effectively and sensibly. And also do this with other functions.

File: app_test.go (hunk 8 above)
Comment (Pham-Gia-Huong) [reply app_test.go]: blank line is space with function

File: app.go [hunk 13]
```diff
@@ -1000,3 +1054,79 @@ func (app *App) Fields() (map[string]*FieldInfo, error) {
...
+	}
+	body, err := parseResponse(response)
+	if err != nil {
+		return nil, err
+	}
+	result, err := decodeCursor(body)
+	return result, nil
+}
+
+//Delete a cursorID
```
Comment (josh-vo) [inline app.go]: Could you please correct comment. It is delete cursor not id, is it?

File: app_test.go [hunk 14]
```diff
@@ -171,6 +171,56 @@ func TestDeleteRecord(t *testing.T) {
...
+
+	cursor := app.createCursorForTest()
+	_, err := app.GetRecordsByCursor(string(cursor.Id))
+
+	if err != nil {
+		t.Errorf("TestGetCursor is failed: %v", err)
+	}
+}
```
Comment (josh-vo) [inline app_test.go]: Could you please return a records and check records size?

File: cursor.go [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+	TotalCount string `json:"totalCount"`
+}
+
+func decodeCursor(b []byte) (c *Cursor, err error) {
+	err = json.Unmarshal(b, &c)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
```
Comment (josh-vo) [inline cursor.go]: Could you please add comment and unit test for this function? 
Thank you.

File: cursor.go [hunk 16]
```diff
@@ -0,0 +1,18 @@
+package kintone
+
+import (
+	"encoding/json"
+)
+
+type Cursor struct {
+	Id         string `json:"id"`
+	TotalCount string `json:"totalCount"`
+}
```
Comment (josh-vo) [inline cursor.go]: Could you please add comment?

File: cursor.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+	}
+	err = json.Unmarshal(b, &t)
+	if err != nil {
+		return nil, err
+	}
+	listRecord, err := DecodeRecords(b)
+	if err != nil {
+		return nil, err
+	}
+	records := &RecordCursor{Records: listRecord, Next: t.next}
```
Comment (josh-vo) [inline cursor.go]: Please correct the name of variable.

File: app.go [hunk 18]
```diff
@@ -1000,3 +1055,79 @@ func (app *App) Fields() (map[string]*FieldInfo, error) {
...
+	}
+	response, err := app.do(request)
+	if err != nil {
+		return nil, err
+	}
+	data, err := parseResponse(response)
+	if err != nil {
+		return nil, err
+	}
+	record, _ := DecodeRecordCursor(data)
```
Comment (josh-vo) [inline app.go]: Please correct the name of variable.

File: app.go [hunk 19]
```diff
@@ -1000,3 +1055,79 @@ func (app *App) Fields() (map[string]*FieldInfo, error) {
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//Using Cursor Id to get all records
+//GetRecordsByCursor return the meta data of the Record in this application
+func (app *App) GetRecordsByCursor(id string) (*RecordCursor, error) {
```
[the remark this answers is not in the document]
Comment (Pham-Gia-Huong) [reply app.go]: Still keep GetRecordsByCursor name

**Pull request merged**